### DISCUSSION TITLE
Link tools with orthogonal router

### DISCUSSION THREAD

**Discussion by lbueschel** [clientIO/joint#1592] [Q&A] [answered]
Hi,

I have run into a few issues with the link tools and after struggling with them for a few days, I hope you can help me. I've created a demo based on the Hello World tutorial, [and you can find my source code in this fiddle](https://jsfiddle.net/lukasbueschel/g0ksv6bh/4/).

![1](https://user-images.githubusercontent.com/49986331/148096889-591034fb-f20e-42c2-9275-481922cab33d.png)

### Issue 1

I don't want to provide the vertices tool, I only want to use the segments tool. Because the link has no vertices, the segment handles are not available. Okay, to continue, I add a vertex to the link.

![2](https://user-images.githubusercontent.com/49986331/148097313-75b868c0-a2dd-4bae-a564-12261517e6a3.png)

The problem is: When moving around the white ellipse, the result is a bit unintended:

![3](https://user-images.githubusercontent.com/49986331/148097563-7135c880-4881-4558-84c4-0b44aa462b45.png)

I know why this happens, no problem. But this is what I want to get:

![Element 1](https://user-images.githubusercontent.com/49986331/148097716-e41a34c0-bc0c-418a-a827-1c1b5d061609.png)

I would like to have...
* the segment handles for the segments of the link calculated by the orthogonal router
* but without having this kind of "fixed" vertices.

Is is possible to implement this behavior with JointJS?

### Issue 2

When moving the handle of the vertical segment to the right, the following cases will occur:
(1) broken
![4](https://user-images.githubusercontent.com/49986331/148098354-412d27e3-e1a2-4029-ae86-b87f7859c156.png)

(2) okay
![5](https://user-images.githubusercontent.com/49986331/148098379-db66b6ed-4fb3-46ae-9d79-5833a09a77b1.png)

(3) unexpected
![6](https://user-images.githubusercontent.com/49986331/148098411-8416b3d4-a0d5-4d80-87c2-3e2144c382fd.png)

(4) okay
![7](https://user-images.githubusercontent.com/49986331/148098432-9ea12242-0522-4e57-ad8b-fb4f1b0ba047.png)

So the first case is quite annoying. Is there anything I can do to fix this?

The third case is unexpected for me because I thought that this is what the [padding option of the router](https://resources.jointjs.com/docs/jointjs/v3.4/joint.html#routers.orthogonal) is for:

```javascript
link.router('orthogonal', {
  padding: 0
});
```

I tried to explain my confusion in this issue #1589 but maybe I am misunderstanding the documentation or using the wrong options or maybe the options of the different components are blocking each other...

Furthermore the behavior isn't deterministic:

![8](https://user-images.githubusercontent.com/49986331/148103813-de9e92e7-5a07-4201-b752-8a95faae0b67.png)

![9](https://user-images.githubusercontent.com/49986331/148103826-703c3869-6011-49f6-8489-8f8427dfbe92.png)

 Do you have any hints for me? I would be very happy to have your help!

Thanks in advance,
Lukas

**Comment by kumilingus** [comment, marked as answer]
The shape of the link is controlled by several attributes (source, target, anchors, connection points, router, connector, vertices). To simplify it let's assume it is only:

- position of the source element
- position of the target element
- vertices
- router

The orthogonal router finds the route automatically based on the source and target position and defined vertices. If you don't want to use vertices, what is the segment tool supposed to change in order to modify the link's shape?

By design, the segment tool is not meant to be used with a router other than `normal`.

Now assume you somehow change the shape of the link with the segment tool. What do you want to happen when the user moves one of the elements? I don't understand it from the picture? Do you want the shape of the link to be reset to the default orthogonal link shape? Or somehow move only the closest waypoint.

The issue with `padding=0` might be an edge case bug. Can you create a JSFiddle with such a setup?

**Reply by lbueschel** [reply]
Thank you very much for your answer.

Indeed, the most important hint is that the segment tool is for the `normal` router only.

Yes, I wanted to move the closest waypoint, but that intention seems to be obsolete there now.

Here you can find a JSFiddle for the padding problem: https://jsfiddle.net/lukasbueschel/g0ksv6bh/4/ (I already posted the link above).

Best regards
Lukas

**Comment by nt294** [comment]
Hi, I believe I am experiencing this exact same issue. I would like to use the orthogonal router and use a handle to be able to drag the links. Did you ever manage to come to a solution for this?

**Reply by kumilingus** [reply]
See https://github.com/clientIO/joint/issues/2044#issuecomment-1429675458.

**Reply by nt294** [reply]
Thanks for your response. Is it possible with this library to implement this behaviour? i.e. dragging the link to update the route

**Reply by kumilingus** [reply]
As comment says:
> If you want to use segment tools with orthogonal like router you can check out this [CodePen](https://codepen.io/jointjs/pen/ExRxwQK/a2dac907b335b0a8c770bbbbf9a3a121). Introduces a custom element view that translates near vertices when moving.
